app/scripts/controllers/service.js

```diff
@@ -60,6 +61,50 @@ angular.module('openshiftConsole')
           }
         );
 
+        watches.push(DataService.watch("services", context, function(services) {
+          $scope.services = services.by("metadata.name");
+        }));
+
+        watches.push(DataService.watch("pods", context, function(pods) {
+          var service = $scope.service;
+          if (!service) {
+            DataService
+              .get("services", $routeParams.service, context)
+              .then(function(service) {
+                getPodsForService(service, pods);
+              });
+              return;
+          }
+          getPodsForService(service, pods);
+        }));
+
+        // receive pods for the current service scope only when the service object is available
+        function getPodsForService(service, pods) {
+          $scope.podsForService = [];
+          var ls = new LabelSelector(service.spec.selector);
+          angular.forEach(pods.by("metadata.name"), function(pod) {
+            if (ls.matches(pod)) {
+              $scope.podsForService.push(pod);
+            }
+          });
+        }
+
+        watches.push(DataService.watch("endpoints", context, function(endpoints) {
+          $scope.podsWithEndpoints = {};
+          angular.forEach(endpoints.by("metadata.name"), function(endpoint) {
```

Comment:
endpoints.by("metadata.name") returns a map of objects indexed by the metadata name, so iterating over that map is wasting cycles, you can do a direct lookup in that map using $routeParams.service.

To follow our code style use _.each when looping over either arrays or hashes.

You don't need to do undefined checks when trying to find JSON subpaths if you use _.get, see the usage below to get targetRef.Kind

```
var svcEndpoint = endpoints.by("metadata.name")[$routeParams.service];
if (svcEndpoint) {
  _.each(svcEndpoint.subsets, function(subset) {
    _.each(subset.addresses, function(address) {
      if (_.get(address, "targetRef.Kind") === "Pod") {
        $scope.podsWithEndpoints[address.targetRef.name] = true;
      }
    });
  });
}
```